File: MoveZeros.js
/**
 * @param {number[]} nums
 * @return {void} Do not return anything, modify nums in-place instead.
 */
var moveZeroes = function (nums) {
	let count = 0;
	for (let i = 0; i < nums.length; i++) {
		if (nums[i] !== 0) {
			nums[count] = nums[i];
			count++;
		}
	}
	while (count < nums.length) {
		nums[count] = 0;
		count++;
	}
};

var moveZeroes = function (nums) {
	let i_n_z = 0;
	let i = 0;
	while (i < nums.length) {
		if (nums[i] !== 0) {
			if (i !== i_n_z) {
				const t = nums[i];
				nums[i] = nums[i_n_z];
				nums[i_n_z] = t;
			}
			i_n_z++;
		}
		i++
	}
};

File: leetcode/moveZeros.py
from typing import List


class Solution:
    def moveZeroes(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        move_zeroes(nums)


def move_zeroes(nums):
    n = len(nums)

    l = 0  # i_most_left_zero
    r = 0  # i

    while r < n:
        if nums[r] != 0:
            if nums[l] == 0:  # l != r
                nums[l], nums[r] = nums[r], nums[l]
            l += 1
        r += 1

        # if nums[l] == 0 and nums[r] == 0:
        #     r += 1
        # elif nums[l] == 0:
        #     nums[l], nums[r] = nums[r], nums[l]
        #     l += 1
        #     r += 1
        # else:
        #     l += 1
        #     r += 1


move_zeroes([0, 1, 2, 0, 3, 4])
